crates/proto/src/rr/rdata/tsig.rs

```diff
@@ -0,0 +1,770 @@
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+           Other Data          /
+///  /                                                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// ```
+pub fn read(decoder: &mut BinDecoder<'_>) -> ProtoResult<TSIG> {
+    let algorithm = Algorithm::read(decoder)?;
+    let time_high = decoder.read_u16()?.unverified(/*valid as any u16*/) as u64;
+    let time_low = decoder.read_u32()?.unverified(/*valid as any u32*/) as u64;
+    let time = (time_high << 32) + time_low;
+    let fudge = decoder.read_u16()?.unverified(/*valid as any u16*/);
+    let mac_size = decoder.read_u16()?.unverified(/* TODO maybe compare to alg out size, but we don't actually know that value? */);
+    let mac =
+        decoder.read_vec(mac_size as usize)?.unverified(/*valid as any vec of the right size*/);
+    let oid = decoder.read_u16()?.unverified(/*valid as any u16*/);
+    let error = decoder.read_u16()?.unverified(/*valid as any u16*/);
+    let other_len = decoder.read_u16()?.unverified(/*valid as any u16*/);
+    let other =
+        decoder.read_vec(other_len as usize)?.unverified(/*valid as any vec ot the right size*/);
+
+    Ok(TSIG {
+        algorithm,
+        time,
+        fudge,
+        mac,
+        oid,
+        error,
+        other,
+    })
+}
+
+/// Write the RData from the given Encoder
+///
+/// ```text
+///                       1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+///   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  /                         Algorithm Name                        /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |                                                               |
+///  |          Time Signed          +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |                               |            Fudge              |
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          MAC Size             |                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+             MAC               /
+///  /                                                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          Original ID          |            Error              |
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          Other Len            |                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+           Other Data          /
+///  /                                                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// ```
+pub fn emit(encoder: &mut BinEncoder<'_>, tsig: &TSIG) -> ProtoResult<()> {
+    tsig.algorithm.emit(encoder)?;
+    encoder.emit_u16((tsig.time >> 32) as u16)?;
```

Comment:
we may want to be a little pedantic on this and use `u16::try_from` on all of these downcast conversions.

Reply:
The assertion in TSIG::new (or the way it's decoded) makes this can't fail, but I guess it could help catch issues if code changes